Update scrollable sizes once per `View:update`

Guldoman opened the pull request:
This reduces the number of `get_[h_]scrollable_size` calls by around three times.

franko commented:
I like this idea, of course this goes in the good direction.

I thought that the fields you are adding and using are just a cache about some UI calculation that we compute at the beginning of the frame and we use through all the UI computations within the frame. These field variables are different from the other object's variables in that they are just temporary cache values only valid during a single frame calculation.

Following this idea I thought that we may make the fact that these variables are just a in-frame cache **explicit** may be putting them in a `self.cache` variable. We may even restore the function self.get_h_scrollbar_size so that it just return the corresponding value info self.cache.

In this was it would be crystal clear that these are not state variables of the node but just cache.

Then I have a question: doesn't this mechanism make the check on the document's change id redundant ?

Guldoman commented:
> Following this idea I thought that we may make the fact that these variables are just a in-frame cache **explicit** may be putting them in a `self.cache` variable.

Yes, I like this idea, I'll work on this.

> We may even restore the function self.get_h_scrollbar_size so that it just return the corresponding value info self.cache.

I don't know if I understood, but I don't think this is a good idea. For example the `scale` plugin expects (after #494 also for the `h_` part) an updated value from `get_[h_]scrollbar_size`. Also keep in mind that I consider #481 as reverted.

> Then I have a question: doesn't this mechanism make the check on the document's change id redundant ?

I believe so, but this PR doesn't really consider the cache added in #481, taking into account that I reverted it in #489 (and I tested this mainly with #481 reverted). So ideally both this and #489 should be merged.

franko commented:
> I don't know if I understood, but I don't think this is a good idea. For example the `scale` plugin expects (after #494 also for the `h_` part) an updated value from `get_[h_]scrollbar_size`. Also keep in mind that I consider #481 as reverted.

AFAIK #481 was never reverted. Did I miss something ? I just made some comments about merging it too quickly to let me have a look.

For your point about the scale plugin I don't understand your point. Can you expand a little bit more please ?

In my mind it is okay because the caching of UI computation is only done within a frame and in this time lapse there will be no change in the state of the application. The application change state only between one frame the the following one.

franko commented:
I think I understood: #489 revert #481 and this PR is supposed to be done on top of #489, right ?

Guldoman commented:
> I think I understood: #489 revert #481 and this PR is supposed to be done on top of #489, right ?

It's not an hard dependency, but yes.

> For your point about the scale plugin I don't understand your point. Can you expand a little bit more please ?

The `scale` plugin requests the scrollable sizes before and after scaling fonts **in the same "frame"**, so having a cached value will report the wrong sizes on the second call.

franko commented:
> The `scale` plugin requests the scrollable sizes before and after scaling fonts **in the same "frame"**, so having a cached value will report the wrong sizes on the second call.

Ok, you are right, I see now. The scale plugin operate between one frame and the following one, when the application's state can change. It is special because, while it's not in the phase where we render the UI (in the frame) it still need to have access to UI computations to restore the scroll's position of the view.

We may do that the `self.cache` is set to nil when the rendering of the frame is done. Functions like the scale plugin that tries to use `self.get_scrollable_size` function or its h counterpart will get an up to date value because the function will see the cache is not set and will compute the value and return it.

So the function `self.get_scrollable_size` and the h one needs to check if self.cache is not nil and use the cached value in this case or perform the actual computation otherwise.

For the self.cache variable we just need to ensure it is set at the beginning of the frame and unset at the end.

Guldoman commented:
> We may do that the `self.cache` is set to nil when the rendering of the frame is done. Functions like the scale plugin that tries to use `self.get_scrollable_size` function or its h counterpart will get an up to date value because the function will see the cache is not set and will compute the value and return it.
> 
> So the function `self.get_scrollable_size` and the h one needs to check if self.cache is not nil and use the cached value in this case or perform the actual computation otherwise.
> 
> For the self.cache variable we just need to ensure it is set at the beginning of the frame and unset at the end.

This seems reasonable.

Only problem is I found out that `Treeview` defines its own `cache` that conflicts with this one :sweat_smile:.
I'll see if I can make it work without changing name.

franko commented:
> Only problem is I found out that `Treeview` defines its own `cache` that conflicts with this one 😅.
> I'll see if I can make it work without changing name.

We may call the new cache `draw_cache` or something like this because it is related to the phase when we draw the UI using methods called 'draw'.

The cache of the TreeView is persisted across frames and is not about UI drawing but about the listing of the file and directory entries that should be displayed.

Guldoman commented:
> We may do that the `self.cache` is set to nil when the rendering of the frame is done. Functions like the scale plugin that tries to use `self.get_scrollable_size` function or its h counterpart will get an up to date value because the function will see the cache is not set and will compute the value and return it.

Ok, that doesn't work because `scale` calls `get_scrollable_size` twice in a row: during the first call the cache gets set, so with the second call the cached result is returned. A `View:invalidate_step_cache` is needed between the two calls.

Also I started to call it `step_cache` because I think that it should be available during the entire step (`update`+`draw`), but should we invalidate the cache after `update` **and** after `draw`? If not, what should we do with views that aren't the active view in a node so they don't get the `draw` call?

franko commented:
> Ok, that doesn't work because `scale` calls `get_scrollable_size` twice in a row: during the first call the cache gets set, so with the second call the cached result is returned. A `View:invalidate_step_cache` is needed between the two calls.

I meant that the cache get set only at the beginning of the frame, during the update method. Otherwise, when not set the cache remains off, it is not automatically updated.

> Also I started to call it `step_cache` because I think that it should be available during the entire step (`update`+`draw`), but should we invalidate the cache after `update` **and** after `draw`? If not, what should we do with views that aren't the active view in a node so they don't get the `draw` call?

To me the cache should be invalidated only after the "draw" methods.

Guldoman commented:
What should we do with events? They happen before `update` so the cache is still disabled at that point. Should the `step_cache` be enabled on some events like `on_mouse_moved`?

Guldoman commented:
> Should the `step_cache` be enabled on some events like `on_mouse_moved`?

For now I enabled it only for `on_mouse_moved`, tell me if this should be changed.

Also notice that I reverted #481 here too, because without the revert this commit would have been weird.
I'll rebase #489 if this gets merged first.

I moved the cache from `View` to `DocView` as it gets used only there.

franko commented:
Sorry for my misguided suggestion, as you find out the drawing cache values are needed also when handling events.

I come up with this:

https://github.com/lite-xl/lite-xl/tree/scroll_size_update_once_2

which is basically a variant of your implementation except we keep the cache activated and we update the cache explicitly. It wort of work but I have an important concern.

The conclusion I drawn is: **caching the 'h_scrollable_size' within a frame is not enough** because this call is really, really expensive. I have made some tests using a file with long lines and it was sending the CPU usage to the roof even caching the value at the beginning of the frame.

So I guess we are left with three possible options:

1. cache this information at the document level
2. completely remove the horizontal scrolling bar implementation
3. implement the approximative solution I was suggesting (do like if only code_font is used and compute the h_scrollable_size as a multiple of a single letter's width for this font)

For solution (1) I don't like it because it leaks UI drawing stuff into the document's object. Normally they should be decorrelated and we may end-up with a solution that works but it would probably be inelegant and fragile.

I suggest to explore solution (3) in a branch but I am open to suggestions.

Otherwise we may consider this as a challenge for the application's architecture. Having an horizontal scrollbar is a very reasonable thing to have and the current architecture is making this really difficult. Maybe we can come up with a good architectural solution to overcome the problem and that could be beneficial for the project to implement new features, not only the horizontal scrollbar.

Guldoman commented:
> I come up with this:
> 
> https://github.com/lite-xl/lite-xl/tree/scroll_size_update_once_2
> 
> which is basically a variant of your implementation except we keep the cache activated and we update the cache explicitly. It wort of work but I have an important concern.

I'll check it out.

> 
> The conclusion I drawn is: **caching the 'h_scrollable_size' within a frame is not enough** because this call is really, really expensive. I have made some tests using a file with long lines and it was sending the CPU usage to the roof even caching the value at the beginning of the frame.

Have you tried with #489? It improves performance quite a bit.

> 1. cache this information at the document level

> For solution (1) I don't like it because it leaks UI drawing stuff into the document's object. Normally they should be decorrelated and we may end-up with a solution that works but it would probably be inelegant and fragile.

We could provide a line length calculation function to `Doc` from `DocView` (like `get_col_x_offset`) and get a precise (not just counting characters, but counting the actual pixel sizes) max line length, thus avoiding having to recalculate it in `DocView` as the value in `doc.long_lines.length` would already be there. The only problem with that is that we still need to invalidate it after a font change (and also try avoiding entire document re-calculations).

> 3. implement the approximative solution I was suggesting (do like if only code_font is used and compute the h_scrollable_size as a multiple of a single letter's width for this font)

> I suggest to explore solution (3) in a branch but I am open to suggestions.

This should be an optimization for both `DocView:get_col_x_offset` and `DocView:get_x_offset_col`. I'll explore this possibility.

Guldoman commented:
> > 3. implement the approximative solution I was suggesting (do like if only code_font is used and compute the h_scrollable_size as a multiple of a single letter's width for this font)
> 
> > I suggest to explore solution (3) in a branch but I am open to suggestions.
> 
> This should be an optimization for both `DocView:get_col_x_offset` and `DocView:get_x_offset_col`. I'll explore this possibility.
> 
> EDIT: Would it be possible to add a `renderer.font` api call to check if a specified font is monospace?

I made some tests with #489 applied and this doesn't yield much optimization. That's because even with a monospace font, we still need to count how many `tab` characters are present and consider them in the calculation (as `tab`s can have a different width from other characters). If we still want to optimize this, we'll have to do it in `src/renderer.c:ren_get_font_width`.
With #489, I think that most of the time is spent in the tokenizer.

Could we use a cache in `Highlighter`? I think that this would be reasonable as it only depends on the state of the `Doc` and not the font.

franko commented:
I am now a little bit lost, I will try to summarize.

What I understood is that the document-based cache system as done in #481 is required because the call to compute the h scrollable size is too expensive to be done even if only once per frame. In other term we need to use a cache for h_scrollable_size and invalidate the cache only when the document has some changes.

On the other side #481 while working it is not complete and requires some modifications to account for config or scale changes. Is this right ? We already have the needed modifications to make #481 work ?

You are also mentioning @Guldoman that the approximative computation I was suggesting is too expensive because of tabs but we may fix this in the C side in the renderer. Is this right ?$

@Guldoman could you summarize also from your side please so that we check we are aligned ?

Guldoman commented:
Because it was discussed that we should keep the "immediate-mode" as much as possible, this last days I worked on reducing the weight that `get_col_x_offset` brings. This resulted in #489 and in trying to speculate the size based on the font being monospace.
I believe #489 is a good solution, as we were doing something as slow as calculating the x offset character by character.
For the monospace speculation, doing this in lua can be demanding as we still have to check for `tab`s and their size. But doing it in C might be fast enough.

> What I understood is that the document-based cache system as done in #481 is required because the call to compute the h scrollable size is too expensive to be done even if only once per frame. In other term we need to use a cache for h_scrollable_size and invalidate the cache only when the document has some changes.

> On the other side #481 while working it is not complete and requires some modifications to account for config or scale changes. Is this right ?

Yes, you are correct.

> We already have the needed modifications to make #481 work ?

No. We need to find a way to invalidate this cache when fonts change, also keeping in mind that plugins can change it too.

> You are also mentioning @Guldoman that the approximative computation I was suggesting is too expensive because of tabs but we may fix this in the C side in the renderer. Is this right ?$

Yes, in the next few days I'll try to see if this is doable.

Also what I was initially trying to do in this PR, was checking which superfluous or repeated calculation we were doing each frame.

franko commented:
> Because it was discussed that we should keep the "immediate-mode" as much as possible, this last days I worked on reducing the weight that `get_col_x_offset` brings. This resulted in #489 and in trying to speculate the size based on the font being monospace.
> I believe #489 is a good solution, as we were doing something as slow as calculating the x offset character by character.
> For the monospace speculation, doing this in lua can be demanding as we still have to check for `tab`s and their size. But doing it in C might be fast enough.

Okay, I agree with that and I am also very favorable to improve this method which is beneficial in any case. I am also favorable to a C implementation.

The problem with #489 is that it is also reverting #481 while they are mostly unrelated. I propose to not revert #481 in #489 and to merge in master.

> > We already have the needed modifications to make #481 work ?
> 
> No. We need to find a way to invalidate this cache when fonts change, also keeping in mind that plugins can change it too.

I had another idea about this. I thought we may use for the docview a "presentation hash". The presentation hash should include the document content's state (change id) but also all the setting that affect the geometric layout of the "presentation" of the content. So we should include which font is used and at which size but for example not the color of the style.

With this change #481 may became complete because we would use the "presentation hash" instead of the change id of the document so any change in style would be "seen". We have just to be careful about all the things that needs to be included in the presentation hash.

> > You are also mentioning @Guldoman that the approximative computation I was suggesting is too expensive because of tabs but we may fix this in the C side in the renderer. Is this right ?$
> 
> Yes, in the next few days I'll try to see if this is doable.

Still worth exploring I think even if with #489 and #481 with the presentation hash we may already have a performing solution.

> Also what I was initially trying to do in this PR, was checking which superfluous or repeated calculation we were doing each frame.

That's a good thing, there are many small things that can be optimized and you already pointed out a number of them.

franko commented:
It seems this PR is a dead end and the improvements needed for the horizontal scrolling bar are probably in the continuation of #489 and #481. I propose to close this PR if you agree.

To go further, as you know, I already removed the horizontal scrollbar from the fix-2.0.2 branch. I propose that we remove also the horizontal scrolling bar also from the master branch and we continue the work in a topic branch.

You are doing a really good job but the horizontal scrolling bar turned out to be difficult to implement in relation to the immediate-mode nature of Lite XL UI. I think it is better to work in a separate branch the remove the "pressure" form the master branch while continuing to work in a topic branch.

If you agree we can revert the commit one by one in reverse order or revert all the related changes in one big commit. I may help with the git gymnastics if needed.

Guldoman commented:
> The problem with #489 is that it is also reverting #481 while they are mostly unrelated. I propose to not revert #481 in #489 and to merge in master.

All right, I'll update the PR.

> I had another idea about this. I thought we may use for the docview a "presentation hash". The presentation hash should include the document content's state (change id) but also all the setting that affect the geometric layout of the "presentation" of the content. So we should include which font is used and at which size but for example not the color of the style.
> 
> With this change #481 may became complete because we would use the "presentation hash" instead of the change id of the document so any change in style would be "seen". We have just to be careful about all the things that needs to be included in the presentation hash.

Yeah, this seems the way to go.

> 
> > > You are also mentioning @Guldoman that the approximative computation I was suggesting is too expensive because of tabs but we may fix this in the C side in the renderer. Is this right ?$
> > 
> > 
> > Yes, in the next few days I'll try to see if this is doable.
> 
> Still worth exploring I think even if with #489 and #481 with the presentation hash we may already have a performing solution.

I have some initial code, I'll test if the improvements are worth it.

> It seems this PR is a dead end and the improvements needed for the horizontal scrolling bar are probably in the continuation of #489 and #481. I propose to close this PR if you agree.
> 
> To go further, as you know, I already removed the horizontal scrollbar from the fix-2.0.2 branch. I propose that we remove also the horizontal scrolling bar also from the master branch and we continue the work in a topic branch.

> If you agree we can revert the commit one by one in reverse order or revert all the related changes in one big commit. I may help with the git gymnastics if needed.

Thats fine for me. Revert PR is incoming.

Should I open a new branch on this repository or should I keep it on my fork?

franko commented:
> Should I open a new branch on this repository or should I keep it on my fork?

In the lite-xl main repository, please.
